src/components/ModelingSidebar/ModelingPanes/FeatureTreePane.tsx

```diff
@@ -0,0 +1,383 @@
+      >
+        {kclManager.isExecuting ? (
+          <Loading>Building feature tree...</Loading>
+        ) : (
+          <>
+            {parseErrors.length > 0 && (
+              <div
+                className={`absolute inset-0 rounded-lg p-2 ${
+                  operationList.length &&
+                  `bg-destroy-10/40 dark:bg-destroy-80/40`
+                }`}
+              >
+                <div className="text-sm bg-destroy-80 text-chalkboard-10 py-1 px-2 rounded flex gap-2 items-center">
+                  <p className="flex-1">
+                    Errors found in KCL code.
+                    <br />
+                    Please fix them before continuing.
+                  </p>
+                  <button
+                    onClick={goToError}
+                    className="bg-chalkboard-10 text-destroy-80 p-1 rounded-sm flex-none hover:bg-chalkboard-10 hover:border-destroy-70 hover:text-destroy-80 border-transparent"
+                  >
+                    View error
+                  </button>
+                </div>
+              </div>
+            )}
+            {operationList.map((operation) => {
+              const key = `${operation.type}-${
+                'name' in operation ? operation.name : 'anonymous'
+              }-${
+                'sourceRange' in operation ? operation.sourceRange[0] : 'start'
+              }`
+
+              return <OperationItem key={key} item={operation} />
+            })}
+          </>
+        )}
+      </section>
+    </div>
+  )
+}
+
+export const visibilityMap = new Map<string, boolean>()
+
+interface VisibilityToggleProps {
+  entityId: string
+  initialVisibility: boolean
+  onVisibilityChange?: () => void
+}
+
+/**
+ * A button that toggles the visibility of an entity
+ * tied to an artifact in the feature tree.
+ * TODO: this is unimplemented and will be used for
+ * default planes after we fix them and add them to the artifact graph / feature tree
+ */
+const VisibilityToggle = (props: VisibilityToggleProps) => {
+  const [visible, setVisible] = useState(props.initialVisibility)
+
+  function handleToggleVisible() {
+    setVisible(!visible)
+    visibilityMap.set(props.entityId, !visible)
+    props.onVisibilityChange?.()
+  }
+
+  return (
+    <button
+      onClick={handleToggleVisible}
+      className="border-transparent p-0 m-0"
+    >
+      <CustomIcon
+        name={visible ? 'eyeOpen' : 'eyeCrossedOut'}
+        className={`w-5 h-5 ${
+          visible
+            ? 'hidden group-hover/item:block group-focus-within/item:block'
+            : 'text-chalkboard-50'
+        }`}
+      />
+    </button>
+  )
+}
+
+/**
+ * Wait until a predicate is true or a timeout is reached
+ * TODO: this is a temporary solution until we have a better way to
+ * wait for things like selections to be set. It's whack get rid of it if you're reading this.
+ */
+function pollUntil(timeout: number, predicate: () => boolean, interval = 20) {
```

Comment:
Why do we need this?

Reply:
In the case of `openCodePane()`, which is called by `openToFunctionDefinition()`, this should be implemented with a state machine. As it is now, it creates new race conditions.

Reply:
Yeah sure I'd love that. My hypothesis is that "set selection" on the modelingMachine is an XState action with async work, so there is no way for us to send "set selection" followed immediately by any other modelingMachine event without something like this at the moment. I saw the "set selection" race condition as a standalone issue that I didn't want to roll into this large of a PR. But I support that!